linkerd/proxy/transport/src/listen.rs

```diff
@@ -79,68 +63,42 @@ impl<A: OrigDstAddr> Bind<A> {
     pub fn keepalive(&self) -> Option<Duration> {
         self.keepalive
     }
-}
-
-impl<O: OrigDstAddr> listen::Bind for Bind<O> {
-    type Connection = Connection;
-    type Listen = Listen<O>;
-
-    fn bind(&self) -> std::io::Result<Listen<O>> {
-        let tcp = std::net::TcpListener::bind(self.bind_addr)?;
-        let listen_addr = tcp.local_addr()?;
-        Ok(Listen {
-            listen_addr,
-            keepalive: self.keepalive,
-            orig_dst_addr: self.orig_dst_addr.clone(),
-            state: State::Init(Some(tcp)),
-        })
-    }
-}
 
-impl<O> listen::Listen for Listen<O>
-where
-    O: OrigDstAddr,
-{
-    type Connection = Connection;
-    type Error = std::io::Error;
-
-    fn listen_addr(&self) -> SocketAddr {
-        self.listen_addr
-    }
+    pub fn bind<S: Future>(
+        &self,
+        shutdown: S,
+    ) -> std::io::Result<(SocketAddr, impl Stream<Item = io::Result<Connection>>)> {
+        let listen = std::net::TcpListener::bind(self.bind_addr)?;
+        let addr = listen.local_addr()?;
+        let keepalive = self.keepalive;
+        let get_orig = self.orig_dst_addr.clone();
+
+        let accept = try_stream! {
+            // The tokio listener is built lazily so that it is initialized on
+            // the proper runtime.
+            let listen = tokio::net::TcpListener::from_std(listen).expect("listener must be valid");
+
+            futures::pin_mut!(listen);
+            futures::pin_mut!(shutdown);
+            loop {
+                let accept = tokio::select! {
```

Comment:
TIOLI: Is this where the `select!`/`async_stream` incompatibility was worked around? It _might_ be worth documenting that for future generations....

Reply:
I was previously trying to yield out of the select, which is not supported by async-stream. I don't think that should be documented here, though. It wouldn't help to make this code easier to undertand...